dotnet/Utils/RichPath.cs

```diff
@@ -0,0 +1,439 @@
+            }
+
+            return memoryStream;
+        }
+
+        private string CachedFilePath => 
+            System.IO.Path.Combine(
+                cacheLocation,
+                containerName,
+                Path.Replace('/', System.IO.Path.DirectorySeparatorChar) // Note that this translates a blob path (/ separators) to a local path (potentially using \)
+                );
+
+        private LocalPath CacheFileLocally()
+        {
+            var etagPath = CachedFilePath + ".etag";
+
+            string storedEtag = null;
+            if (File.Exists(etagPath))
+            {
+                storedEtag = File.ReadAllText(etagPath);
+            }
+
+            var tmpPath = CachedFilePath + Guid.NewGuid();
+            try
+            {
+                Directory.CreateDirectory(System.IO.Path.GetDirectoryName(CachedFilePath));
+                // Download blob to a tmp file, to avoid overwriting a potentially existing copy:
+                var blobReference = blobContainerClient.GetBlobReference(Path);
+                blobReference.DownloadToFile(
+                    tmpPath,
+                    FileMode.Create,
+                    new AccessCondition { IfNoneMatchETag = storedEtag });
+                // If download succeeded (no exception), remove old file if it exists and put new file in place:
+                if (File.Exists(CachedFilePath)) {
+                    File.Delete(CachedFilePath);
+                }
+                File.Move(tmpPath, CachedFilePath);
+                File.WriteAllText(etagPath, blobReference.Properties.ETag);
+            }
+            catch (StorageException e)
+            {
+                if (e.RequestInformation.HttpStatusCode != 304) // 304 ~ Not modified, i.e., we are fine
+                {
+                    throw;
+                }
+            }
+            finally
+            { 
+                if (File.Exists(tmpPath))
+                {
+                    File.Delete(tmpPath);
+                }
+            }
+
+            return new LocalPath(CachedFilePath);
+        }
+
+        public override Stream ReadAsBinaryPotentiallyUncompress()
+        {
+            if (cacheLocation == null)
+            {
+                return ReadDirectlyFromBlobStoragePotentiallyUncompress();
+            }
+            return CacheFileLocally().ReadAsBinaryPotentiallyUncompress();
+        }
+
+        public override void Store(object data)
+        {
+            // Create new file that has same extension as this file...
+            var tempFile = System.IO.Path.Combine(
+                System.IO.Path.GetTempPath(),
+                Guid.NewGuid().ToString() + Path.Split('/').Last());
+            try
+            {
+                var tempLocalPath = new LocalPath(tempFile);
+                tempLocalPath.Store(data);
+                var blobReference = blobContainerClient.GetBlockBlobReference(Path);
+                blobReference.UploadFromFile(tempFile);
```

Comment:
This is a minor point, but it may improve rare performance issues.

When the file is uploaded, the cache is not updated (if any). I would suspect that the `blobReference` updates the `etag` on upload and the cache (+etag) can be updated at this point. Otherwise, the next time that the file gets read, a download would need to happen (assuming we have a cache).

Reply:
Done.